Referencing a variable in stack right before pattern matching results in a crash.

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 12.3.1, 21E258
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.1.6 (clang-1316.0.21.2)

getwd: /Users/batuhan/Documents/V/learn-v
vmodules: /Users/batuhan/.vmodules
vroot: /Users/batuhan/Documents/V/v-lang
vexe: /Users/batuhan/Documents/V/v-lang/v
vexe mtime: 2022-05-01 11:12:00
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 ec92d46.aed2d0c

Git version: git version 2.32.0 (Apple Git-132)
Git vroot status: weekly.2022.16-117-gaed2d0ca
.git/config present: true
thirdparty/tcc status: thirdparty-macos-arm64 173c526e

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg main.v`
```v
import os


struct Player {
mut:
	x int
	y int
	level int
}

struct Enemy {
mut:
	x int
	y int
	damage f64
}


type PlayerOrEnemy = Player | Enemy


fn main() {

	mut entity := PlayerOrEnemy(Player{10, 12, 3})

	x_move := os.input('X movement: ').int()
	y_move := os.input('Y movement: ').int()

	mut ref := &entity

	match mut entity {
		Player {
			entity.x += x_move
			entity.y += y_move

			println('Player is moved to $entity.x, $entity.y and its level is $entity.level')
		}
		Enemy {
			entity.x += x_move
			entity.y += y_move

			println('Enemy is moved to $entity.x, $entity.y and its damage is $entity.damage')
		}
	}
}
```

**What did you expect to see?**

I was expecting to see the same output before I added the line which says 'mut ref := &entity'.


**What did you see instead?**
```
main.v:29:6: warning: unused variable: `ref`
   27 |     y_move := os.input('Y movement: ').int()
   28 | 
   29 |     mut ref := &entity
      |         ~~~
   30 | 
   31 |     match mut entity {
==================
   ^~~~~~~~~~~~~~~~
/tmp/v_501/main.4259885740918205918.tmp.c:15787:11: error: member reference type 'main__PlayerOrEnemy *' (aka 'struct main__PlayerOrEnemy *') is a pointer; did you mean to use '->'?
                (*entity._main__Player).x += x_move;
                  ~~~~~~^
                        ->
/tmp/v_501/main.4259885740918205918.tmp.c:15788:11: error: member reference type 'main__PlayerOrEnemy *' (aka 'struct main__PlayerOrEnemy *') is a pointer; did you mean to use '->'?
                (*entity._main__Player).y += y_move;
                  ~~~~~~^
                        ->
/tmp/v_501/main.4259885740918205918.tmp.c:15789:114: error: member reference type 'main__PlayerOrEnemy *' (aka 'struct main__PlayerOrEnemy *') is a pointer; did you mean to use '->'?
                println( str_intp(4, _MOV((StrIntpData[]){{_SLIT("Player is moved to "), /*100 &int*/0xfe07, {.d_i32 = (*entity._main__Player).x}}, {_SLIT(", "), /*100 &int*/0xfe07, {.d_i32 = (*entity._main__Player).y}}, {_SLIT(" and its level is "), /*100 &int*/0xfe07, {.d_i32 = (*entity._main__Player).level}}, {_SLIT0, 0, { .d_c = 0 }}})));
                                                                                                                         ~~~~~~^
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
Error occurred right after I added the line
```v
mut ref := &entity
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3656,6 +3656,7 @@ pub fn (mut c Checker) mark_as_referenced(mut node ast.Expr, as_interface bool) 
 								node.obj.is_auto_heap = true
 							}
 						}
+						.sum_type, .interface_ {}
 						else {
 							node.obj.is_auto_heap = true
 						}

--- a/vlib/v/tests/match_reference_sumtype_var_test.v
+++ b/vlib/v/tests/match_reference_sumtype_var_test.v
@@ -1,0 +1,42 @@
+struct Player {
+mut:
+	x     int
+	y     int
+	level int
+}
+
+struct Enemy {
+mut:
+	x      int
+	y      int
+	damage f64
+}
+
+type PlayerOrEnemy = Enemy | Player
+
+fn test_match_reference_sumtype_var() {
+	mut entity := PlayerOrEnemy(Player{10, 12, 3})
+
+	x_move := 11
+	y_move := 22
+
+	mut ref := &entity
+
+	match mut entity {
+		Player {
+			entity.x += x_move
+			entity.y += y_move
+
+			println('Player is moved to $entity.x, $entity.y and its level is $entity.level')
+		}
+		Enemy {
+			entity.x += x_move
+			entity.y += y_move
+
+			println('Enemy is moved to $entity.x, $entity.y and its damage is $entity.damage')
+		}
+	}
+
+	println(typeof(ref).name)
+	assert true
+}

